File: 백준/Gold/2110. 공유기 설치/공유기 설치.cc
#include <iostream>
#include <algorithm>
#include <vector>
using namespace std;

int cntRouter(int dist, vector<int>& router){
    int cnt = 1;
    int cur = router[0];
    for (int i=1; i<router.size(); i++){
        if (router[i] - cur >= dist){
            cnt++;
            cur = router[i];
        }
    }
    return cnt;
}

int binarySearch(int left, int right, int target, vector<int>& router){
    while(left <= right){
        int mid = (left + right) / 2;
        int installed = cntRouter(mid, router);
        if (installed >= target) {
            left = mid + 1;
        } else {
            right = mid - 1;
        }
    }
    return left - 1;
}

int main(){
    int n, c;
    cin >> n >> c;
    vector<int> house(n, 0);
    for (int i=0; i<n; i++) {
        cin >> house[i];
    }
    sort(house.begin(), house.end());
    cout << binarySearch(1, house[n-1]-house[0], c, house);
    return 0;
}

File: 백준/Gold/2110. 공유기 설치/공유기 설치.py
import sys
input = sys.stdin.readline

def countRouter(dist, house):
	cnt = 1
	cur = house[0]
	for h in house:
		if (h - cur) >= dist:
			cur = h
			cnt += 1
	return cnt

def binarySearch(start, end, target, house):
	while start <= end:
		mid = (start + end) // 2
		cnt = countRouter(mid, house)
		if cnt >= target:
			start = mid + 1
		else:
			end = mid - 1
	return start - 1

n, c = map(int, input().split())
house = [int(input()) for _ in range(n)]
house.sort()
print(binarySearch(1, house[-1] - house[0], c, house))